### ISSUE TITLE
Add production dependency to adress precision and floating point issues

### ISSUE THREAD

**Issue by nicolashefti** [formulajs/formulajs#126]
Kyle1297 is proposing to add decimal.js as a production dependency to Formula.js #123 

If accepted, this should be generalized in the whole library (to avoid having a new dependency for 1 function). This could be added the 2023 roadmap.

**Pro**
- Solve former issues #112 #109
- Healthy and active repo https://www.npmjs.com/package/decimal.js

**Cons**
- Bundle size (+20%)
- Too much functionalities for what we actually need ?

Before:
-rw-r--r--  1 nicolashefti  staff  346107 13 Dec 18:17 formula.js
-rw-r--r--  1 nicolashefti  staff  136196 13 Dec 18:17 formula.min.js

After:
-rw-r--r--  1 nicolashefti  staff  415244 13 Dec 18:19 formula.js // +20%
-rw-r--r--  1 nicolashefti  staff  170066 13 Dec 18:19 formula.min.js // +25%

@pbadenski @bornova , what do you think?

**Comment by nicolashefti** [comment]
Other possibilities :
- adding just what's necessary to the Formula.js utils to adresse locally the issue
- Big.js? Other dependencies?

**Comment by Kyle1297** [comment]
Thanks for doing the full breakdown of this @nicolashefti. 

Another library recommended by DecimalJS itself is [decimalJS-light](https://github.com/MikeMcl/decimal.js-light/). It doesn't look like it is actively worked on that much anymore, but it is significantly smaller while giving us what we need. 

BigJS could be a good option, but we would need to make sure it works like we want it to since it does things a little differently. Here is what DecimalJS has to say about it: 
"The library is similar to [bignumber.js](https://github.com/MikeMcl/bignumber.js/), but here precision is specified in terms of significant digits rather than decimal places, and all calculations are rounded to the precision (similar to Python's decimal module) rather than just those involving division."

Another alternative is minifying the decimalJS code. See here: https://www.npmjs.com/package/decimal.js#minify.

Of course, we could also just implement this ourselves... but it feels like we would be wasting our time on building a solution that already exists elsewhere.

**Comment by nicolashefti** [comment]
- [decimal.js-light](https://github.com/MikeMcl/decimal.js-light) does not seem to be highly maintained (or the implementation is finished...)
- Minification : DecimalJs would be bundled and minified within FormulaJs. This is arlready the case for jStats and Bessel (the two other dependencies).

@bornova, I know you just started a new job. Do you have a few minutes to share your opinion on this?

**Comment by bornova** [comment]
Hi everyone, although I am not completely opposed to using decimal.js to address some of the rounding issues, I am not completely sold on it either.  Some of my favorite math libraries, such as math.js, uses decimal.js to address floating point calculation issues so it definitely has it uses.  However, we should remember Excel itself also has these floating errors.  Excel just happens to use precision of 15, if I remember correctly, to hide some of these issues.  For example, try doing `A1:1.2` `B1:1.1` and `C1=A1-A2` then set the precision to 16.  You will see the incorrect result there too.

https://learn.microsoft.com/en-us/office/troubleshoot/excel/floating-point-arithmetic-inaccurate-result

Also, if we implement decimal.js, does that mean we have to start parsing expressions in arguments passed to functions? For example, in the newly suggested TEXT function, if we try the following test case:
```javascript
text.TEXT(1.2 - 1.1 + 1.1555, '########.##########%').should.equal('125.55%')
```
it will give you an incorrect answer, even with decimal.js, unless we parsed the value argument as
```javascript
new Decimal(1.2).minus(1.1).plus(1.1555)
```

(@Kyle1297 please don't take this as a negative comment.  What you have proposed for this new TEXT function is quite remarkable and I certainly appreciate the work you have put in it!)

In summary, I am not so sure if it makes sense to introduce another 3rd party library (great as it may be) to address a few edge cases with formula.js

**Comment by nicolashefti** [comment]
In a way, I like the idea of saying: there a floatting point issue and you should know about it. We don't hide it wrapping all the calculations with an additional precision library.

If you want don't like what you see, then ROUND your numbers (that's Method 1 proposed on the help page of Excel).

**Comment by zyw** [comment]
@nicolashefti When can this function be completed?

**Comment by nicolashefti** [comment]
@zyw the PR #123 has been closed without being merged. The discussion here goes into the direction of _not_ dealing with floating point issues inside Formula.js.

Feel free to reopen the discussion or add your point of view.
